In [1]:
%reload_ext nb_black

import warnings

warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import SGD

<IPython.core.display.Javascript object>

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()
input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype("float32")
X_test = X_test.astype("float32")
X_train /= 255
X_test /= 255

y_train = to_categorical(y_train, nb_classes)
y_test = to_categorical(y_test, nb_classes)

<IPython.core.display.Javascript object>

# ReLU activation and SGD loss

In [3]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))

<IPython.core.display.Javascript object>

In [4]:
model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

<IPython.core.display.Javascript object>

In [5]:
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 0s 997us/step - loss: 1.3426 - accuracy: 0.6532
Epoch 2/20
469/469 [==============================] - 0s 986us/step - loss: 0.5257 - accuracy: 0.8659
Epoch 3/20
469/469 [==============================] - 0s 909us/step - loss: 0.3951 - accuracy: 0.8909
Epoch 4/20
469/469 [==============================] - 0s 923us/step - loss: 0.3456 - accuracy: 0.9021
Epoch 5/20
469/469 [==============================] - 0s 869us/step - loss: 0.3163 - accuracy: 0.9097
Epoch 6/20
469/469 [==============================] - 0s 857us/step - loss: 0.2953 - accuracy: 0.9149
Epoch 7/20
469/469 [==============================] - 0s 870us/step - loss: 0.2787 - accuracy: 0.9198
Epoch 8/20
469/469 [==============================] - 0s 855us/step - loss: 0.2644 - accuracy: 0.9240
Epoch 9/20
469/469 [==============================] - 0s 872us/step - loss: 0.2522 - accuracy: 0.9272
Epoch 10/20
469/469 [==============================] - 0s 887us/step - loss: 0.241

<IPython.core.display.Javascript object>

In [6]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.17308348417282104
Test accuracy: 0.9490000009536743


<IPython.core.display.Javascript object>

In [7]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 0s 879us/step - loss: 1.9842 - accuracy: 0.3344
Epoch 2/20
469/469 [==============================] - 0s 859us/step - loss: 0.8775 - accuracy: 0.7729
Epoch 3/20
469/469 [==============================] - 0s 894us/step - loss: 0.4757 - accuracy: 0.8629
Epoch 4/20
469/469 [==============================] - 0s 917us/step - loss: 0.3774 - accuracy: 0.8907
Epoch 5/20
469/469 [==============================] - 0s 956us/step - loss: 0.3247 - accuracy: 0.9062
Epoch 6/20
469/469 [==============================] - 0s 902us/step - loss: 0.2882 - accuracy: 0.9167
Epoch 7/20
469/469 [==============================] - 0s 887us/step - loss: 0.2604 - accuracy: 0.9247
Epoch 8/20
469/469 [==============================] - 0s 903us/step - loss: 0.2377 - accuracy: 0.9311
Epoch 9/20
469/469 [==============================] - 0s 871us/step - loss: 0.2188 - accuracy: 0.9361
Epoch 10/20
469/469 [==============================] - 0s 893us/step - loss: 0.203

<IPython.core.display.Javascript object>

In [8]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.12963488698005676
Test accuracy: 0.9606000185012817


<IPython.core.display.Javascript object>

In [9]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 3ms/step - loss: 1.8797 - accuracy: 0.3439
Epoch 2/20
469/469 [==============================] - 1s 3ms/step - loss: 0.6186 - accuracy: 0.8279
Epoch 3/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3632 - accuracy: 0.8964
Epoch 4/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2813 - accuracy: 0.9185
Epoch 5/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2323 - accuracy: 0.9328
Epoch 6/20
469/469 [==============================] - 1s 3ms/step - loss: 0.1999 - accuracy: 0.9419
Epoch 7/20
469/469 [==============================] - 1s 3ms/step - loss: 0.1766 - accuracy: 0.9486
Epoch 8/20
469/469 [==============================] - 1s 3ms/step - loss: 0.1580 - accuracy: 0.9538
Epoch 9/20
469/469 [==============================] - 1s 3ms/step - loss: 0.1441 - accuracy: 0.9577
Epoch 10/20
469/469 [==============================] - 1s 3ms/step - loss: 0.1307 - accuracy: 0.9623

<IPython.core.display.Javascript object>

In [10]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09538017213344574
Test accuracy: 0.9706000089645386


<IPython.core.display.Javascript object>

In [11]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(1024, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 9s 18ms/step - loss: 1.5713 - accuracy: 0.5212
Epoch 2/20
469/469 [==============================] - 9s 19ms/step - loss: 0.5210 - accuracy: 0.8482
Epoch 3/20
469/469 [==============================] - 12s 25ms/step - loss: 0.3147 - accuracy: 0.9124
Epoch 4/20
469/469 [==============================] - 9s 18ms/step - loss: 0.2283 - accuracy: 0.9372
Epoch 5/20
469/469 [==============================] - 12s 25ms/step - loss: 0.1905 - accuracy: 0.9473
Epoch 6/20
469/469 [==============================] - 10s 20ms/step - loss: 0.1468 - accuracy: 0.9591
Epoch 7/20
469/469 [==============================] - 12s 26ms/step - loss: 0.1234 - accuracy: 0.9654
Epoch 8/20
469/469 [==============================] - 12s 26ms/step - loss: 0.1058 - accuracy: 0.9700
Epoch 9/20
469/469 [==============================] - 12s 25ms/step - loss: 0.0908 - accuracy: 0.9743
Epoch 10/20
469/469 [==============================] - 12s 26ms/step - loss: 0.0780 -

<IPython.core.display.Javascript object>

In [12]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09052375704050064
Test accuracy: 0.9750000238418579


<IPython.core.display.Javascript object>

The most accurate model came from using 9 layers. The speed of the model was much slower. The run time of a 7 layer model was noticeably faster but performed with a worse accuracy.

# Tanh activation and SGD loss

In [13]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(10, activation="softmax"))

<IPython.core.display.Javascript object>

In [14]:
model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

<IPython.core.display.Javascript object>

In [15]:
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 1.0217 - accuracy: 0.7430
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 0.5210 - accuracy: 0.8669
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 0.4213 - accuracy: 0.8878
Epoch 4/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3732 - accuracy: 0.8979
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3435 - accuracy: 0.9040
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3224 - accuracy: 0.9093
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3063 - accuracy: 0.9133
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2932 - accuracy: 0.9166
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2820 - accuracy: 0.9195
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2722 - accuracy: 0.9219

<IPython.core.display.Javascript object>

In [16]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.20599287748336792
Test accuracy: 0.9412999749183655


<IPython.core.display.Javascript object>

In [17]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 1.4077 - accuracy: 0.6426
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 0.7767 - accuracy: 0.8385
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 0.5562 - accuracy: 0.8761
Epoch 4/20
469/469 [==============================] - 1s 1ms/step - loss: 0.4479 - accuracy: 0.8934
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3844 - accuracy: 0.9047
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3427 - accuracy: 0.9125
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 0.3123 - accuracy: 0.9184
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2886 - accuracy: 0.9232
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2695 - accuracy: 0.9278
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 0.2531 - accuracy: 0.9316

<IPython.core.display.Javascript object>

In [18]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.16711029410362244
Test accuracy: 0.951200008392334


<IPython.core.display.Javascript object>

In [19]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(256, activation="tanh"))
model.add(Dense(128, activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.1680 - accuracy: 0.7130
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.6266 - accuracy: 0.8693
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 0.4549 - accuracy: 0.8958
Epoch 4/20
469/469 [==============================] - 3s 6ms/step - loss: 0.3694 - accuracy: 0.9117
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 0.3170 - accuracy: 0.9204
Epoch 6/20
469/469 [==============================] - 3s 6ms/step - loss: 0.2806 - accuracy: 0.9277
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2522 - accuracy: 0.9345
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2292 - accuracy: 0.9399
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2098 - accuracy: 0.9441
Epoch 10/20
469/469 [==============================] - 2s 5ms/step - loss: 0.1933 - accuracy: 0.9484

<IPython.core.display.Javascript object>

In [20]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.120052270591259
Test accuracy: 0.9656999707221985


<IPython.core.display.Javascript object>

In [21]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="tanh"))
model.add(Dense(1024, activation="tanh"))
model.add(Dense(512, activation="tanh"))
model.add(Dense(256, activation="tanh"))
model.add(Dense(128, activation="tanh"))
model.add(Dense(64, activation="tanh"))
model.add(Dense(32, activation="tanh"))
model.add(Dense(16, activation="tanh"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 15s 33ms/step - loss: 0.9837 - accuracy: 0.8138
Epoch 2/20
469/469 [==============================] - 15s 31ms/step - loss: 0.5038 - accuracy: 0.9003
Epoch 3/20
469/469 [==============================] - 15s 32ms/step - loss: 0.3632 - accuracy: 0.9192 ETA: 0s - loss: 0.3632 - accuracy: 0.91
Epoch 4/20
469/469 [==============================] - 15s 32ms/step - loss: 0.2946 - accuracy: 0.9307
Epoch 5/20
469/469 [==============================] - 15s 33ms/step - loss: 0.2507 - accuracy: 0.9395
Epoch 6/20
469/469 [==============================] - 15s 33ms/step - loss: 0.2194 - accuracy: 0.94620s - loss: 0.2196 - accuracy: 0.94 - ETA: 0s - loss: 0.2194 - accuracy: 0.94
Epoch 7/20
469/469 [==============================] - 16s 33ms/step - loss: 0.1945 - accuracy: 0.9513
Epoch 8/20
469/469 [==============================] - 15s 33ms/step - loss: 0.1749 - accuracy: 0.9561
Epoch 9/20
469/469 [==============================] - 15s 33ms/step 

<IPython.core.display.Javascript object>

In [22]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.0916651263833046
Test accuracy: 0.9750000238418579


<IPython.core.display.Javascript object>

The model with 9 layers had the best test accuracy, however it was noticeably slower. Additionally, the accuracy was only better by around 0.01. Using tanh activation provided very similar results and the accuracy was just barely lower than when using ReLU.

# Sigmoid activation and SGD loss

In [23]:
# 3 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])
model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2932 - accuracy: 0.1537
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2374 - accuracy: 0.3281
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 2.1802 - accuracy: 0.4724
Epoch 4/20
469/469 [==============================] - 1s 1ms/step - loss: 2.0997 - accuracy: 0.5575
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 1.9830 - accuracy: 0.5992
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 1.8244 - accuracy: 0.6241
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 1.6387 - accuracy: 0.6496
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 1.4552 - accuracy: 0.6811
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 1.2945 - accuracy: 0.7051
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 1.1597 - accuracy: 0.7312

<IPython.core.display.Javascript object>

In [24]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.5553178787231445
Test accuracy: 0.8575999736785889


<IPython.core.display.Javascript object>

In [25]:
# 5 Layers
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3219 - accuracy: 0.1036
Epoch 2/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3011 - accuracy: 0.1124
Epoch 3/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3008 - accuracy: 0.1124
Epoch 4/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3006 - accuracy: 0.1124
Epoch 5/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3004 - accuracy: 0.1124
Epoch 6/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3002 - accuracy: 0.1124
Epoch 7/20
469/469 [==============================] - 1s 1ms/step - loss: 2.3000 - accuracy: 0.1124
Epoch 8/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2997 - accuracy: 0.1124
Epoch 9/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2995 - accuracy: 0.1124
Epoch 10/20
469/469 [==============================] - 1s 1ms/step - loss: 2.2993 - accuracy: 0.1124

<IPython.core.display.Javascript object>

In [26]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 2.2956039905548096
Test accuracy: 0.11349999904632568


<IPython.core.display.Javascript object>

In [27]:
# 7 Layers
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(256, activation="sigmoid"))
model.add(Dense(128, activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3239 - accuracy: 0.1018
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3015 - accuracy: 0.1102
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3012 - accuracy: 0.1124
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 10/20
469/469 [==============================] - 3s 5ms/step - loss: 2.3013 - accuracy: 0.1124

<IPython.core.display.Javascript object>

In [28]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 2.3009867668151855
Test accuracy: 0.11349999904632568


<IPython.core.display.Javascript object>

In [29]:
# 9 Layers
model = Sequential()
model.add(Dense(2048, input_dim=X_train.shape[1], activation="sigmoid"))
model.add(Dense(1024, activation="sigmoid"))
model.add(Dense(512, activation="sigmoid"))
model.add(Dense(256, activation="sigmoid"))
model.add(Dense(128, activation="sigmoid"))
model.add(Dense(64, activation="sigmoid"))
model.add(Dense(32, activation="sigmoid"))
model.add(Dense(16, activation="sigmoid"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 15s 32ms/step - loss: 2.3438 - accuracy: 0.0979
Epoch 2/20
469/469 [==============================] - 15s 32ms/step - loss: 2.3016 - accuracy: 0.1124
Epoch 3/20
469/469 [==============================] - 15s 31ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 4/20
469/469 [==============================] - 15s 31ms/step - loss: 2.3013 - accuracy: 0.11240s - loss: 2.3013 - accuracy: 0.
Epoch 5/20
469/469 [==============================] - 17s 36ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 6/20
469/469 [==============================] - 16s 35ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 7/20
469/469 [==============================] - 16s 34ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 8/20
469/469 [==============================] - 16s 35ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 9/20
469/469 [==============================] - 16s 33ms/step - loss: 2.3013 - accuracy: 0.1124
Epoch 10/20
469/469 [=============================

<IPython.core.display.Javascript object>

In [30]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 2.301093578338623
Test accuracy: 0.11349999904632568


<IPython.core.display.Javascript object>

The test accuracy for the sigmoid tests are all pretty low. It appears using sigmoid activation is not a good idea for this model.

# ReLU activation and SGD loss Changing Batch Sizes

In [31]:
# Using 64 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 3s 3ms/step - loss: 1.3685 - accuracy: 0.5548
Epoch 2/20
938/938 [==============================] - 3s 3ms/step - loss: 0.3357 - accuracy: 0.9037
Epoch 3/20
938/938 [==============================] - 3s 4ms/step - loss: 0.2133 - accuracy: 0.9384
Epoch 4/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1626 - accuracy: 0.9521
Epoch 5/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1323 - accuracy: 0.9617
Epoch 6/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1134 - accuracy: 0.9669
Epoch 7/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0966 - accuracy: 0.9714
Epoch 8/20
938/938 [==============================] - 5s 5ms/step - loss: 0.0833 - accuracy: 0.9752
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0729 - accuracy: 0.9786
Epoch 10/20
938/938 [==============================] - 4s 4ms/step - loss: 0.0637 - accuracy: 0.9817

<IPython.core.display.Javascript object>

In [32]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09109868854284286
Test accuracy: 0.9761000275611877


<IPython.core.display.Javascript object>

In [33]:
# Using 32 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=32, epochs=20, verbose=1)

Epoch 1/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.8199 - accuracy: 0.7394
Epoch 2/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.2208 - accuracy: 0.9354: 0s - loss:
Epoch 3/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1434 - accuracy: 0.9584
Epoch 4/20
1875/1875 [==============================] - 5s 2ms/step - loss: 0.1077 - accuracy: 0.9689
Epoch 5/20
1875/1875 [==============================] - 5s 2ms/step - loss: 0.0849 - accuracy: 0.9746
Epoch 6/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0690 - accuracy: 0.9794
Epoch 7/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0552 - accuracy: 0.9844
Epoch 8/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0449 - accuracy: 0.9869
Epoch 9/20
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0354 - accuracy: 0.9901
Epoch 10/20
1875/1875 [==============================] - 5s 3ms/step -

<IPython.core.display.Javascript object>

In [34]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09937319904565811
Test accuracy: 0.9793000221252441


<IPython.core.display.Javascript object>

In [35]:
# Using 16 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=16, epochs=20, verbose=1)

Epoch 1/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.5334 - accuracy: 0.8378
Epoch 2/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.1491 - accuracy: 0.9564
Epoch 3/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0973 - accuracy: 0.9717
Epoch 4/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0702 - accuracy: 0.9794
Epoch 5/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0512 - accuracy: 0.9845
Epoch 6/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0378 - accuracy: 0.9883
Epoch 7/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0269 - accuracy: 0.9920
Epoch 8/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0203 - accuracy: 0.9941
Epoch 9/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.0155 - accuracy: 0.9956
Epoch 10/20
3750/3750 [==============================] - 8s 2ms/step - loss: 0.011

<IPython.core.display.Javascript object>

In [36]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09268534183502197
Test accuracy: 0.9811000227928162


<IPython.core.display.Javascript object>

In [37]:
# Using 8 as a batch size
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=8, epochs=20, verbose=1)

Epoch 1/20
7500/7500 [==============================] - 21s 3ms/step - loss: 0.3793 - accuracy: 0.8821
Epoch 2/20
7500/7500 [==============================] - 20s 3ms/step - loss: 0.1132 - accuracy: 0.9654
Epoch 3/20
7500/7500 [==============================] - 19s 2ms/step - loss: 0.0740 - accuracy: 0.9771
Epoch 4/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0510 - accuracy: 0.9837
Epoch 5/20
7500/7500 [==============================] - 17s 2ms/step - loss: 0.0377 - accuracy: 0.9880
Epoch 6/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0281 - accuracy: 0.9913
Epoch 7/20
7500/7500 [==============================] - 17s 2ms/step - loss: 0.0225 - accuracy: 0.9924
Epoch 8/20
7500/7500 [==============================] - 17s 2ms/step - loss: 0.0181 - accuracy: 0.9940
Epoch 9/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0127 - accuracy: 0.9960
Epoch 10/20
7500/7500 [==============================] - 15s 2ms/step - l

<IPython.core.display.Javascript object>

In [38]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.08209272474050522
Test accuracy: 0.983299970626831


<IPython.core.display.Javascript object>

The score increases as the batch size decreases, however the run time increases as well. It was more noticeable when the batch size got to 16 and lower. The best score was obtained from using a batch size of 8, but it took more than twice the time. 

Choosing between batch sizes, I would recommend using a batch size of 64 because it provides an accuracy that is almost just as good as having a batch size of 8 but also executing in less than half the time it took for the 8 batch size model to run.

# ReLU Activation, SGD Loss, Batchsize of 64, and Changing the Amount of Neurons 

In [39]:
# Having constant amount of 32 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(32, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 1s 1ms/step - loss: 1.9277 - accuracy: 0.3392
Epoch 2/20
938/938 [==============================] - 1s 1ms/step - loss: 0.5727 - accuracy: 0.8259
Epoch 3/20
938/938 [==============================] - 1s 1ms/step - loss: 0.3653 - accuracy: 0.8937
Epoch 4/20
938/938 [==============================] - 1s 1ms/step - loss: 0.2955 - accuracy: 0.9143
Epoch 5/20
938/938 [==============================] - 1s 1ms/step - loss: 0.2500 - accuracy: 0.9272
Epoch 6/20
938/938 [==============================] - 2s 2ms/step - loss: 0.2186 - accuracy: 0.9359
Epoch 7/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1950 - accuracy: 0.9421
Epoch 8/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1766 - accuracy: 0.9478
Epoch 9/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1630 - accuracy: 0.9517
Epoch 10/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1512 - accuracy: 0.9549

<IPython.core.display.Javascript object>

In [40]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.11589862406253815
Test accuracy: 0.9639999866485596


<IPython.core.display.Javascript object>

In [41]:
# Having constant amount of 64 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 1s 1ms/step - loss: 1.5969 - accuracy: 0.4852
Epoch 2/20
938/938 [==============================] - 1s 1ms/step - loss: 0.4878 - accuracy: 0.8548
Epoch 3/20
938/938 [==============================] - 1s 1ms/step - loss: 0.3130 - accuracy: 0.9082
Epoch 4/20
938/938 [==============================] - 1s 1ms/step - loss: 0.2424 - accuracy: 0.9278
Epoch 5/20
938/938 [==============================] - 1s 1ms/step - loss: 0.2037 - accuracy: 0.9391
Epoch 6/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1761 - accuracy: 0.9472
Epoch 7/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1569 - accuracy: 0.9527
Epoch 8/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1433 - accuracy: 0.9567
Epoch 9/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1311 - accuracy: 0.9605
Epoch 10/20
938/938 [==============================] - 1s 1ms/step - loss: 0.1217 - accuracy: 0.9632

<IPython.core.display.Javascript object>

In [42]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.10847555845975876
Test accuracy: 0.9666000008583069


<IPython.core.display.Javascript object>

In [43]:
# Having constant amount of 128 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(128, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 2s 2ms/step - loss: 1.2429 - accuracy: 0.6127
Epoch 2/20
938/938 [==============================] - 2s 2ms/step - loss: 0.3570 - accuracy: 0.8931
Epoch 3/20
938/938 [==============================] - 2s 2ms/step - loss: 0.2519 - accuracy: 0.9258
Epoch 4/20
938/938 [==============================] - 2s 2ms/step - loss: 0.2030 - accuracy: 0.9404
Epoch 5/20
938/938 [==============================] - 2s 2ms/step - loss: 0.1703 - accuracy: 0.9495
Epoch 6/20
938/938 [==============================] - 2s 2ms/step - loss: 0.1486 - accuracy: 0.9564
Epoch 7/20
938/938 [==============================] - 2s 2ms/step - loss: 0.1324 - accuracy: 0.9611
Epoch 8/20
938/938 [==============================] - 2s 2ms/step - loss: 0.1188 - accuracy: 0.9647
Epoch 9/20
938/938 [==============================] - 2s 2ms/step - loss: 0.1068 - accuracy: 0.9688
Epoch 10/20
938/938 [==============================] - 2s 2ms/step - loss: 0.0974 - accuracy: 0.9719

<IPython.core.display.Javascript object>

In [44]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09070105850696564
Test accuracy: 0.972100019454956


<IPython.core.display.Javascript object>

In [45]:
# Having constant amount of 256 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(256, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 3s 3ms/step - loss: 1.1523 - accuracy: 0.6581
Epoch 2/20
938/938 [==============================] - 3s 3ms/step - loss: 0.3200 - accuracy: 0.9050
Epoch 3/20
938/938 [==============================] - 3s 3ms/step - loss: 0.2359 - accuracy: 0.9292
Epoch 4/20
938/938 [==============================] - 3s 4ms/step - loss: 0.1884 - accuracy: 0.9439
Epoch 5/20
938/938 [==============================] - 3s 4ms/step - loss: 0.1572 - accuracy: 0.9524
Epoch 6/20
938/938 [==============================] - 3s 4ms/step - loss: 0.1357 - accuracy: 0.9597
Epoch 7/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1177 - accuracy: 0.9652
Epoch 8/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1037 - accuracy: 0.9692
Epoch 9/20
938/938 [==============================] - 3s 3ms/step - loss: 0.0919 - accuracy: 0.9724
Epoch 10/20
938/938 [==============================] - 3s 3ms/step - loss: 0.0821 - accuracy: 0.9755

<IPython.core.display.Javascript object>

In [46]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.07917606085538864
Test accuracy: 0.9771000146865845


<IPython.core.display.Javascript object>

In [47]:
# Having constant amount of 512 neurons throughout entire model until output layer
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(512, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 14s 15ms/step - loss: 1.0357 - accuracy: 0.6840
Epoch 2/20
938/938 [==============================] - 14s 15ms/step - loss: 0.2944 - accuracy: 0.91400s - l
Epoch 3/20
938/938 [==============================] - 15s 16ms/step - loss: 0.2129 - accuracy: 0.9380
Epoch 4/20
938/938 [==============================] - 14s 15ms/step - loss: 0.1684 - accuracy: 0.95080s - los
Epoch 5/20
938/938 [==============================] - 14s 15ms/step - loss: 0.1390 - accuracy: 0.9587
Epoch 6/20
938/938 [==============================] - 14s 15ms/step - loss: 0.1184 - accuracy: 0.9649
Epoch 7/20
938/938 [==============================] - 14s 15ms/step - loss: 0.1008 - accuracy: 0.9695
Epoch 8/20
938/938 [==============================] - 15s 16ms/step - loss: 0.0876 - accuracy: 0.9744
Epoch 9/20
938/938 [==============================] - 17s 18ms/step - loss: 0.0777 - accuracy: 0.9770
Epoch 10/20
938/938 [==============================] - 15s 16ms/step

<IPython.core.display.Javascript object>

In [48]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.07709846645593643
Test accuracy: 0.9779000282287598


<IPython.core.display.Javascript object>

In [49]:
# Having neurons decrease throughout model
model = Sequential()
model.add(Dense(512, input_dim=X_train.shape[1], activation="relu"))
model.add(Dense(256, activation="relu"))
model.add(Dense(128, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(32, activation="relu"))
model.add(Dense(16, activation="relu"))
model.add(Dense(10, activation="softmax"))

model.compile(optimizer="sgd", loss="categorical_crossentropy", metrics=["accuracy"])

model.fit(X_train, y_train, batch_size=64, epochs=20, verbose=1)

Epoch 1/20
938/938 [==============================] - 3s 3ms/step - loss: 1.1634 - accuracy: 0.6368
Epoch 2/20
938/938 [==============================] - 3s 3ms/step - loss: 0.3047 - accuracy: 0.9124
Epoch 3/20
938/938 [==============================] - 3s 3ms/step - loss: 0.2047 - accuracy: 0.9413
Epoch 4/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1609 - accuracy: 0.9535
Epoch 5/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1314 - accuracy: 0.9619
Epoch 6/20
938/938 [==============================] - 3s 3ms/step - loss: 0.1098 - accuracy: 0.9688
Epoch 7/20
938/938 [==============================] - 3s 3ms/step - loss: 0.0946 - accuracy: 0.9726
Epoch 8/20
938/938 [==============================] - 3s 3ms/step - loss: 0.0816 - accuracy: 0.9768
Epoch 9/20
938/938 [==============================] - 3s 3ms/step - loss: 0.0706 - accuracy: 0.9800
Epoch 10/20
938/938 [==============================] - 3s 3ms/step - loss: 0.0619 - accuracy: 0.9814

<IPython.core.display.Javascript object>

In [50]:
score = model.evaluate(X_test, y_test, verbose=0)
print("Test score:", score[0])
print("Test accuracy:", score[1])

Test score: 0.09305624663829803
Test accuracy: 0.9743000268936157


<IPython.core.display.Javascript object>

When using constant amounts of neurons the test accuracy increases as the neurons increase, however, once the neurons get too high the run time starts to take longer to execute. The models all appear to have similar accuracies, but the model with a constant amound of 256 neurons performs the best in the most optimal run time. It even has a better accuracy than the model with neurons that decrease within each layer.

# Conclusion

Overall, each model is able to reach a very high accuracy, usually over 0.95. models that achieve an accuracy higher than 0.97 start to take a lot longer to execute, and only have a small increase in accuracy. Additionally, whenever the layers were increased the accuracy would increase, however once the layers got past 7 the accuracy would barely increase and the run time would take much longer. Therefore, I figured the most optimal amount of layers would be 7.

It appears the best activation function was the ReLU activation. I was able to obtain an accuracy score of around 0.977 with a quick run time using a constant amound of 256 neurons, ReLU activation, 7 layers, and SGD loss function. 

### Possible Next Steps

Some next steps that could be done for improving the models would be to change the loss functions (incorporate a hinge loss function instead of categorical). Additionally, I could change the optimizer to something other than sgd (adam).